tests/csapi/public_rooms_test.go

```diff
@@ -0,0 +1,188 @@
+
+	"github.com/matrix-org/complement"
+	"github.com/matrix-org/complement/client"
+	"github.com/matrix-org/complement/helpers"
+	"github.com/matrix-org/complement/match"
+	"github.com/matrix-org/complement/must"
+)
+
+func TestPublicRooms(t *testing.T) {
+	deployment := complement.Deploy(t, 1)
+	defer deployment.Destroy(t)
+
+	t.Run("parallel", func(t *testing.T) {
+		// sytest: Name/topic keys are correct
+		t.Run("Name/topic keys are correct", func(t *testing.T) {
+			t.Parallel()
+			authedClient := deployment.Register(t, "hs1", helpers.RegistrationOpts{})
+
+			// Define room configurations matching the Sytest
+			roomConfigs := []struct {
+				alias string
+				name  string
+				topic string
+			}{
+				{"publicroomalias_no_name", "", ""},
+				{"publicroomalias_with_name", "name_1", ""},
+				{"publicroomalias_with_topic", "", "topic_1"},
+				{"publicroomalias_with_name_topic", "name_2", "topic_2"},
+				{"publicroom_with_unicode_chars_name", "un nom français", ""},
+				{"publicroom_with_unicode_chars_topic", "", "un topic à la française"},
+				{"publicroom_with_unicode_chars_name_topic", "un nom français", "un topic à la française"},
+			}
+
+			// Create all rooms with their configurations
+			createdRooms := make(map[string]struct {
+				roomID string
+				name   string
+				topic  string
+			})
+
+			for _, config := range roomConfigs {
+				roomOptions := map[string]interface{}{
+					"visibility":      "public",
+					"room_alias_name": config.alias,
+				}
+
+				if config.name != "" {
+					roomOptions["name"] = config.name
+				}
+				if config.topic != "" {
+					roomOptions["topic"] = config.topic
+				}
+
+				roomID := authedClient.MustCreateRoom(t, roomOptions)
+				createdRooms[config.alias] = struct {
+					roomID string
+					name   string
+					topic  string
+				}{
+					roomID: roomID,
+					name:   config.name,
+					topic:  config.topic,
+				}
+				t.Logf("Created room %s with alias %s", roomID, config.alias)
+			}
+
+			// Poll /publicRooms until all our rooms appear with correct data
+			authedClient.MustDo(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"},
+				client.WithRetryUntil(15*time.Second, func(res *http.Response) bool {
+					body := must.ParseJSON(t, res.Body)
+
+					must.MatchGJSON(
+						t,
+						body,
+						match.JSONKeyPresent("chunk"),
+						match.JSONKeyTypeEqual("chunk", gjson.JSON),
+					)
+
+					chunk := body.Get("chunk")
+					if !chunk.IsArray() {
+						t.Logf("chunk is not an array")
+						return false
+					}
+
+					// Track which rooms we've correctly found
+					foundRooms := make(map[string]bool)
+
+					// Check each room in the public rooms list
+					for _, roomData := range chunk.Array() {
+						// Verify required keys are present
+						must.MatchGJSON(
+							t,
+							roomData,
+							match.JSONKeyPresent("world_readable"),
+							match.JSONKeyPresent("guest_can_join"),
+							match.JSONKeyPresent("num_joined_members"),
+						)
```

Comment:
This raises and exits the process when it finds something wrong. Whereas the checks below are more lenient and we check all rooms before presenting a result.

We should align one way or the other.

I think it would be nice if we used `err := should.MatchGJSON(jsonResult, matchers...)` and gathered up the errors to present for each expected room at the end.